query.go

```diff
@@ -396,15 +402,42 @@ func (q *query) queryPeer(ctx context.Context, ch chan<- *queryUpdate, p peer.ID
 		ch <- &queryUpdate{cause: p, unreachable: []peer.ID{p}}
 		return
 	}
-
 	queryDuration := time.Since(startQuery)
+	// return if we got a nil response
+	// This happens if the self peer decides to not run the query RPC on the remote peer for some reason
+	if newPeers == nil {
+		ch <- &queryUpdate{cause: p, heard: nil, queried: []peer.ID{p}, queryDuration: queryDuration}
+		return
+	}
 
 	// query successful, try to add to RT
 	q.dht.peerFound(q.dht.ctx, p, true)
 
 	// process new peers
-	saw := []peer.ID{}
-	for _, next := range newPeers {
+	saw := make(map[peer.ID]struct{})
+
+	// process signed peer records
+	for pid, record := range newPeers.signedPeerRecords {
+		if pid == q.dht.self { // don't add self.
+			logger.Debugf("PEERS CLOSER -- worker for: %v found self", p)
+			continue
+		}
+
+		// add any other know addresses for the candidate peer.
+		curInfo := q.dht.peerstore.PeerInfo(pid)
+		curInfo.Addrs = append(curInfo.Addrs, record.Addrs...)
+
+		if q.dht.queryPeerFilter(q.dht, curInfo) {
+			saw[pid] = struct{}{}
+			_, err := q.dht.ca.ConsumePeerRecord(record.Envelope, pstore.TempAddrTTL)
+			if err != nil {
+				logger.Warnf("failed to consume signed record in the peerstore, err=%s", err)
+			}
+		}
+	}
+
+	// process unsigned peers
+	for _, next := range newPeers.unsignedPeers {
```

Comment:
We probably want to explicitly not check unsigned addresses for a peer for which we received signed addresses.